### PR TITLE
PPML FGBoost tutorial draft

### PR DESCRIPTION
**Pull request by Litchilitchy**
@jason-dai  @qiyuangong This PR will continue updating, kindly leave any comment here.

### CODE REVIEW

**Comment by jason-dai** [comment]
Please organize the tutorial as follows:
1. Create a couple of simple but non-trivial projects (e.g., for FGBoost, PyTorch VFL, etc.)
2. Write step-by-step guide that using the project as a walking example (i.e., how the user can build a VFL program using BigDL PPML), with brief description of key concepts

File: docs/readthedocs/source/doc/PPML/VFL/overview.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Vertical Federated Learning
+Vertical Federated Learning (VFL) is a federated machine learning case where two data sets share the same sample ID space but differ in feature space.
```
Comment (qiyuangong) [inline docs/readthedocs/source/doc/PPML/VFL/overview.md:2]: two change to multiple

File: docs/readthedocs/source/doc/PPML/VFL/overview.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
+# Vertical Federated Learning
+Vertical Federated Learning (VFL) is a federated machine learning case where two data sets share the same sample ID space but differ in feature space. 
+
+VFL is supported in BigDL PPML. It allows users to train a federated machine learning model where data features are held by different parties. In BigDL PPML, the following VFL scenarios are supported.
+* **Private Set Intersection**: To get data intersection of different VFL parties.
+* **Neural Network Model**: To train common neural network model with Pytorch or Tensorflow backend across VFL parties.
+* **FGBoost Model**: To train gradient boosted decision tree (GBDT) model across VFL parties.
```
Comment (qiyuangong) [inline docs/readthedocs/source/doc/PPML/VFL/overview.md:7]: vfl -> multi-parites

**Comment by Litchilitchy** [comment]
FGBoost walkthrough updated, note the hyperlinks are not valid for now.

**Comment by jason-dai** [comment]
- Create a project showing how to build an FGBoost application (including both python files and scripts)
- Create a .md file to describe the tutorial:

    ```
    1. Overview
    2. Prepare data
    3. Write client programs - showing code snippets from the above project
    4. Run the programs - step-by-step guide on how to run the programs and get the results; you also need to run the server in SGX
    ```

**Comment by Litchilitchy** [comment]
@jason-dai which directory should this example project locates?

**Comment by jason-dai** [comment]
> @jason-dai which directory should this example project locates?

We can decide it later

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
+# FGBoost Regression Tutorial
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:1]: Tutorial for Vertically Federated XGBoost
1. Key Concepts
1.1 FGBoost
1.2 SGX
1.3 FL Server and Client
3. Build FGBoost client applications
3.1 FL Context
3.2 PSI
3.3 Data processing
3.4 FGBoost instance
3.5 Training
3.6 Save & Load
3.7 Prediction
4. Run FGBoost 
4.1 Start FLServer in SGX
4.2 Start FGBoost clients
4.3 Get results

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+To get the data intersection which the 2 parties can do federated learning, we have to the Private Set Intersection (PSI) algorithm first.
+```python
+from bigdl.ppml.fl.algorithms.psi import PSI
+ids = df_train['Id']
+psi = PSI()
+salt = psi.get_salt()
+psi.upload_set(ids, salt)
+intersection = psi.download_intersection()
+df_train = df_train.ix(intersection) # select the intersection part of training data
+```
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:57]: make this simpler, such as:
```python
intersection = psi.get_intersection(ids)
```

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+# the number of clients in this federated learning application
+clientNum: 2
+```
+### Run Start Script
+Then,
+```
+./ppml/scripts/start-fl-server.sh 
+```
+## Start Federated Client Applications
+To run clients in SGX, follow the instructions [above]().
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:106]: no need to run client in SGX

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+### Run Federated Applications
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:108]: add how to config clients

File: python/ppml/example/fgboost_regression/fgboost_regression_party_1.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    ConverNaNToNAString(train_dataset, features_with_nan_party)        
+    ConverNaNToNAString(test_dataset, features_with_nan_party)
+
+    train_dataset = pd.get_dummies(train_dataset, columns=categorical_features_party)
+    test_dataset = pd.get_dummies(test_dataset, columns=categorical_features_party)
+    every_column_except_y= [col for col in train_dataset.columns if col not in ['SalePrice','Id']]
+    y = train_dataset[['SalePrice']] if 'SalePrice' in train_dataset else None
+    return train_dataset[every_column_except_y], y, test_dataset
+
+
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost_regression_party_1.py:71]: if __name__ == "__main__":

File: python/ppml/example/fgboost_regression/fgboost_regression_party_1.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+fgboost_regression = FGBoostRegression()
+
+# party 1 does not own label, so directly pass all the features
+fgboost_regression.fit(x, feature_columns=x.columns, num_round=15)
+
+fgboost_regression.save_model('/tmp/fgboost_model_1.json')
+loaded = FGBoostRegression.load_model('/tmp/fgboost_model_1.json')
+
+result = fgboost_regression.predict(x_test, feature_columns=x_test.columns)
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost_regression_party_1.py:93]: use loaded model to predict

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 9]
```diff
@@ -0,0 +1,172 @@
+# Tutorial for Vertically Federated XGBoost
+This example provides a step-by-step tutorial of running a FGBoost Regression task with 2 parties.
+## 1. Key Concepts
+### 1.1 FGBoost 
+**FGBoost** stands for Federated Gradient Boosted Tree algorithm. It allows multiple parties to run a federated gradient boosted decision tree application.
+### 1.2 PSI
+**PSI** stands for Private Set Intersection algorithm. It compute the intersection of data from different parties and return the result so that parties know which data they could collaborate with.
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:7]: return the result --> what is the result returned?

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+### 1.3 SGX
+**SGX** stands for [Software Guard Extensions](https://www.intel.com/content/www/us/en/developer/tools/software-guard-extensions/overview.html). It helps protect data in use via unique application isolation technology. The data of application running in SGX would be protected.
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:9]: We should talk about HW/CPU protection

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+
+### 1.4 FL Server and Client
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:11]: Focus on:
(1) FL Server and Client are part of BigDL PPML
(2) A centralized FL Server runs in SGX, aggregates intermediate results of clients, and return aggregated results (loss, gradients)
(3) An FL Client runs on each party, computes intermediate result on local data, and updates partial model based on aggregated results from the FL Server

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+
+To get the data intersection which the 2 parties can do federated learning, we have to the Private Set Intersection (PSI) algorithm first.
+```python
+from bigdl.ppml.fl.algorithms.psi import PSI
+ids = df_train['Id']
+psi = PSI()
+salt = psi.get_salt()
+psi.upload_set(ids, salt)
+intersection = psi.download_intersection()
+df_train = df_train.ix(intersection) # select the intersection part of training data
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:56]: this code is not up to date

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+The code is available in projects, including [Client 1 code](fgboost_regression_party_1.py) and [Client 2 code](fgboost_regression_party_2.py). You could directly start two different terminals are run them respectively to start a federated learning, and the order of start does not matter. Following is the detailed step-by-step tutorial to introduce how the code works.
+
+### 2.1 FL Context
+First, import the package and initilize FL Context.
+```python
+from bigdl.ppml.fl import *
+from bigdl.ppml.fl.algorithms.fgboost_regression import FGBoostRegression
+import pandas as pd
+
+init_fl_context()
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:32]: is there a client id?

Comment (Litchilitchy) [reply python/ppml/example/fgboost_regression/fgboost-tutorial.md:32]: Oh this is missed.. Client ID is only regulated in NN FLServer.. These two are actually two different implementation.. will add soon

Comment (Litchilitchy) [reply python/ppml/example/fgboost_regression/fgboost-tutorial.md:32]: @jason-dai just figured out that FGBoost does not require order info. So do we still need to regulate Client ID?

**Comment by Litchilitchy** [comment]
A update PR is created at #4977

File: python/ppml/example/fgboost_regression/data/house-prices-train-2.csv [hunk 14]
```diff
@@ -0,0 +1,1461 @@
+Id,MSZoning,LotArea,Alley,LandContour,LotConfig,Neighborhood,Condition2,HouseStyle,OverallCond,YearRemodAdd,RoofMatl,Exterior2nd,MasVnrArea,ExterCond,BsmtQual,BsmtExposure,BsmtFinSF1,BsmtFinSF2,TotalBsmtSF,HeatingQC,Electrical,2ndFlrSF,GrLivArea,BsmtHalfBath,HalfBath,KitchenAbvGr,TotRmsAbvGrd,Fireplaces,GarageType,GarageFinish,GarageArea,GarageCond,WoodDeckSF,EnclosedPorch,ScreenPorch,PoolQC,MiscFeature,MoSold,SaleType,SalePrice
```
Comment (qiuxin2012) [inline python/ppml/example/fgboost_regression/data/house-prices-train-2.csv:1]: Can we put house prices data on our project?

File: scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/fl/fgboost/FGBoostModel.scala [hunk 15]
```diff
@@ -265,7 +266,11 @@ abstract class FGBoostModel(continuous: Boolean,
     if (flClient == null) {
       throw new IllegalArgumentException("FLClient not initialized.")
     }
-    flClient.fgbostStub.uploadLabel(gradData.build)
+    val response = flClient.fgbostStub.uploadLabel(gradData.build)
+    if (response.getCode == 1) {
```
Comment (qiuxin2012) [inline scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/fl/fgboost/FGBoostModel.scala:270]: Log4Error.invalidOperationError(response.getCode != 1, response.getResponse)

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 16]
```diff
@@ -0,0 +1,172 @@
...
+
+Party 2:
+```python
+df_test = pd.read_csv('house-prices-test-2')
+result = fgboost_regression.predict(df_test, feature_columns=df_test.columns)
+```
+## 3 Run FGBoost
+FL Server is required before running any federated applications. Check [Start FL Server]() section for details.
+### 3.1 Start FL Server in SGX
+To run server in SGX, pull image from dockerhub
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:110]: remove the SGX content and add an TODO for it

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 17]
```diff
@@ -0,0 +1,172 @@
...
+result = fgboost_regression.predict(df_test, feature_columns=df_test.columns)
+```
+
+Party 2:
+```python
+df_test = pd.read_csv('house-prices-test-2')
+result = fgboost_regression.predict(df_test, feature_columns=df_test.columns)
+```
+## 3 Run FGBoost
+FL Server is required before running any federated applications. Check [Start FL Server]() section for details.
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:108]: running the FL Server and Clients on the same node?

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 18]
```diff
@@ -0,0 +1,172 @@
...
+```
+docker pull intelanalytics/bigdl-ppml-trusted-fl-graphene:2.1.0-SNAPSHOT
+```
+prepare the enclave key
+```
+openssl genrsa -3 -out enclave-key.pem 3072
+```
+// TODO: CHECK
+
+Modify the config file `ppml-conf.yaml`
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:120]: where is the file? explain important parameters.

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 19]
```diff
@@ -0,0 +1,172 @@
...
+
+# the number of clients in this federated learning application
+clientNum: 2
+```
+Then start the FL Server
+```
+./ppml/scripts/start-fl-server.sh 
+```
+### 3.2 Start FGBoost Clients
+Modify the config file `ppml-conf.yaml`
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:133]: where is the file? explain important parameters.

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+Start Client 1:
+```bash
+python fgboost_regression_party_1.py
+```
+Start Client 2:
+```bash
+python fgboost_regression_party_2.py
+```
+### 3.3 Get Results
+The first 5 batches of results are printed
```
Comment (jason-dai) [inline python/ppml/example/fgboost_regression/fgboost-tutorial.md:149]: is this training output?

Comment (Litchilitchy) [reply python/ppml/example/fgboost_regression/fgboost-tutorial.md:149]: prediction output, the print code is in scripts.

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md (hunk 17 above)
Comment (Litchilitchy) [reply python/ppml/example/fgboost_regression/fgboost-tutorial.md:108]: Currently, yes. If no, we have to assume some URL like `192.168.0.1` and set `clientTarget` to it, and user need to do some similar modification before they run this example.

File: python/ppml/example/fgboost_regression/fgboost-tutorial.md (hunk 13 above)
Comment (jason-dai) [reply python/ppml/example/fgboost_regression/fgboost-tutorial.md:32]: We should probably keep the interface consistent?
